src/ray/raylet/scheduling/cluster_resource_data.h

```diff
@@ -27,139 +27,526 @@
 
 namespace ray {
 
-const std::string ResourceEnumToString(PredefinedResources resource);
-
-const PredefinedResources ResourceStringToEnum(const std::string &resource);
+using scheduling::ResourceID;
 
 bool IsPredefinedResource(scheduling::ResourceID resource_id);
 
-/// Helper function to compare two vectors with FixedPoint values.
-bool EqualVectors(const std::vector<FixedPoint> &v1, const std::vector<FixedPoint> &v2);
+/// Represents a set of resources.
+/// NOTE: negative values are valid in this set, while 0 is not. This means if any
+/// resource value is changed to 0, the resource will be removed.
+/// TODO(hchen): This class should be independent with tasks. We should move out the
+/// "requires_object_store_memory_" field, and rename this class to ResourceSet.
+class ResourceRequest {
+ public:
+  /// Construct an empty ResourceRequest.
+  ResourceRequest() : ResourceRequest({}, false) {}
 
-/// Convert a vector of doubles to a vector of resource units.
-std::vector<FixedPoint> VectorDoubleToVectorFixedPoint(const std::vector<double> &vector);
+  /// Construct a ResourceRequest with a given resource map.
+  ResourceRequest(absl::flat_hash_map<ResourceID, FixedPoint> resource_map)
+      : ResourceRequest(resource_map, false){};
 
-/// Convert a vector of resource units to a vector of doubles.
-std::vector<double> VectorFixedPointToVectorDouble(
-    const std::vector<FixedPoint> &vector_fp);
+  ResourceRequest(absl::flat_hash_map<ResourceID, FixedPoint> resource_map,
+                  bool requires_object_store_memory)
+      : requires_object_store_memory_(requires_object_store_memory) {
+    for (int i = 0; i < PredefinedResourcesEnum_MAX; i++) {
+      predefined_resources_.push_back(0);
+    }
```

Comment:
I think we can just do `: predefined_resources_(PredefinedResourcesEnum_MAX), requires_object_store_memory_(requires_object_store_memory) {`?